modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/DataStreamRouterProcessor.java

```diff
@@ -0,0 +1,143 @@
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.ingest.common;
+
+import org.elasticsearch.ingest.AbstractProcessor;
+import org.elasticsearch.ingest.ConfigurationUtils;
+import org.elasticsearch.ingest.IngestDocument;
+import org.elasticsearch.ingest.Processor;
+
+import java.util.Locale;
+import java.util.Map;
+import java.util.Objects;
+
+import static org.elasticsearch.ingest.ConfigurationUtils.newConfigurationException;
+
+public final class DataStreamRouterProcessor extends AbstractProcessor {
+    public static final String TYPE = "data_stream_router";
+
+    private static final String DATA_STREAM_PREFIX = "data_stream.";
+    private static final String DATA_STREAM_TYPE = DATA_STREAM_PREFIX + "type";
+    private static final String DATA_STREAM_DATASET = DATA_STREAM_PREFIX + "dataset";
+    private static final String DATA_STREAM_NAMESPACE = DATA_STREAM_PREFIX + "namespace";
+    private static final char[] DISALLOWED_IN_DATASET = new char[] { '\\', '/', '*', '?', '\"', '<', '>', '|', ' ', ',', '#', ':', '-' };
+    private static final char[] DISALLOWED_IN_NAMESPACE = new char[] { '\\', '/', '*', '?', '\"', '<', '>', '|', ' ', ',', '#', ':' };
+    private static final int MAX_LENGTH = 100;
+    private static final char REPLACEMENT_CHAR = '_';
+    private final String dataset;
+    private final String namespace;
+
+    DataStreamRouterProcessor(String tag, String description, String dataset, String namespace) {
+        super(tag, description);
+        this.dataset = dataset;
+        this.namespace = namespace;
+    }
+
+    private static String sanitizeDataStreamField(String s, char[] disallowedInDataset) {
+        if (s == null) {
+            return null;
+        }
+        s = s.toLowerCase(Locale.ROOT);
+        s = s.substring(0, Math.min(s.length(), MAX_LENGTH));
+        for (char c : disallowedInDataset) {
+            s = s.replace(c, REPLACEMENT_CHAR);
+        }
+        return s;
+    }
+
+    @Override
+    public IngestDocument execute(IngestDocument ingestDocument) throws Exception {
+        final String indexName = ingestDocument.getFieldValue(IngestDocument.Metadata.INDEX.getFieldName(), String.class);
+        final String type;
+        final String datasetFallback;
+        final String namespaceFallback;
+        int indexOfFirstDash = indexName.indexOf('-');
+        String illegalDataStreamNameMessage = "invalid data stream name: ["
+            + indexName
+            + "]; must follow naming scheme <type>-<dataset>-<namespace>";
+        if (indexOfFirstDash < 0) {
+            throw new IllegalArgumentException(illegalDataStreamNameMessage);
+        }
+        type = indexName.substring(0, indexOfFirstDash);
```

Comment:
Lets extract this to a separate function.

Reply:
I've extracted some methods but I'm not sure if it's cleaner now. Having support for multiple return values, like in Go, would come in handy for this.